File: csharp/Behavioral/18-Observer/Observer/AuditLog.cs
namespace Observer;

public class AuditLog : IJobCardObserver
{
    private readonly List<string> _entries = new();

    public void Update(string eventName, JobCard jobCard)
    {
        var entry = $"[{DateTime.Now:HH:mm:ss}] {eventName} — Job #{jobCard.JobId} Status: {jobCard.Status}";
        _entries.Add(entry);
        Console.WriteLine($"  📋 [Audit] {entry}");
    }

    public void PrintLog()
    {
        Console.WriteLine($"\n  [AuditLog] {_entries.Count} event(s) recorded:");
        foreach (var e in _entries)
            Console.WriteLine($"    {e}");
    }
}

File: typescript/Behavioral/18-Observer/Observer/AuditLog.ts
import type { JobCard } from "../Subject/JobCard";
import { IJobCardObserver } from "./IJobCardObserver";

export class AuditLog implements IJobCardObserver {
  private readonly entries: string[] = [];

  update(eventName: string, jobCard: JobCard): void {
    const time = new Date().toTimeString().slice(0, 8);
    const entry = `[${time}] ${eventName} — Job #${jobCard.jobId} Status: ${jobCard.status}`;
    this.entries.push(entry);
    console.log(`  📋 [Audit] ${entry}`);
  }

  printLog(): void {
    console.log(`\n  [AuditLog] ${this.entries.length} event(s) recorded:`);
    for (const e of this.entries) console.log(`    ${e}`);
  }
}
